arch/src/x86_64/interrupts.rs

```diff
@@ -30,29 +28,23 @@ const APIC_MODE_NMI: u32 = 0x4;
 const APIC_MODE_EXTINT: u32 = 0x7;
 
 fn get_klapic_reg(klapic: &kvm_lapic_state, reg_offset: usize) -> u32 {
-    let sliceu8 = unsafe {
-        // This array is only accessed as parts of a u32 word, so interpret it as a u8 array.
-        // Cursors are only readable on arrays of u8, not i8(c_char).
-        &*(&klapic.regs[reg_offset..] as *const [i8] as *const [u8])
+    let reg = {
+        let reg = &klapic.regs[reg_offset..reg_offset + 4];
+        unsafe { slice::from_raw_parts(reg.as_ptr().cast(), reg.len()) }
     };
-    let mut reader = Cursor::new(sliceu8);
-    // Following call can't fail if the offsets defined above are correct.
-    reader
-        .read_u32::<LittleEndian>()
-        .expect("Failed to read klapic register")
+
+    let u8_array = reg.try_into().unwrap();
+
+    u32::from_le_bytes(u8_array)
 }
```

Comment:
It would be nice to also get rid of the unsafe block. Something like this would work:

```
    let mut reg = [0 as u8; 4];
    for (i, val) in reg.iter_mut().enumerate() {
        *val = klapic.regs[reg_offset + i] as u8;
    }

    u32::from_le_bytes(reg)
```

Reply:
You're right, and I guess I should have put a bit more faith in LLVM. :) The iterator approach seems to optimize pretty much the same as with the slice casting.